README.md
# GlucNAc/DateTimeInterval

[![Source Code][badge-source]][source]
[![Latest Version][badge-release]][packagist]
[![PHP Version][badge-php]][php]
[![Pipeline status][badge-pipeline]][pipeline]
[![Coverage Status][badge-coverage]][coverage]
[![Total Downloads][badge-downloads]][downloads]

When you want to compute a delay between two DateTimes, it can be quickly tricky to get the result you are expecting for,
and sometimes you need to perform some calculations, which can lead to incorrect result.

DateTimeInterval is wrapper of DateInterval that provides useful methods for calculating delays. Thanks to this library,
there is no need to compute results returned from DateInterval anymore. Moreover, all of its methods are fully tested,
and are explicitly named and commented.

In that way, you don't have to copy/paste the logic you implemented on another project, just install this library and
enjoy.

## Installation

The preferred method of installation is via [Composer][]. Run the following
command to install the package and add it as a requirement to your project's
`composer.json`:

```bash
composer require glucnac/datetimeinterval
```

## What is the purpose ?

To understand the usefulness of this library, let's try to compute some delays.

### Get delays using date_diff()
```php
// Let's say you want to compute some delays between 2022-07-15 and 2022-08-15.

$firstDate = new DateTimeImmutable('2022-07-15');
$secondDate = new DateTimeImmutable('2022-08-15');

$dateInterval = date_diff($secondDate, $firstDate);
// OR
$dateInterval = $firstDate->diff($secondDate);

// You want to count the days between these two dates. Using DateInterval,
// you notice that two properties are available :

// PhpDoc : Totals number of days in the interval span.
$dateInterval->days;

// PhpDoc : Number of days.
$dateInterval->d;

// Well, the doc of ->d is not really clear, let's see the difference between these two returns

$dateInterval->days; // 30
$dateInterval->d; // 0

// Nice, it seems ->d is a relative count, whereas ->days give an absolute count

// What about getting the number of months then ?

// PhpDoc : number of months
$dateInterval->m // 1

// Seems to be ok. What about changing the year to next one :
$firstDate = new DateTimeImmutable('2022-07-15');
$secondDate = new DateTimeImmutable('2023-08-15');
$dateInterval = date_diff($secondDate, $firstDate);

$dateInterval->m // 1

// ->m also returns a relative count
```

Actually, ->y, ->m, ->d, ->h, ->i, ->s, ->f : all are returning a relative count, which means that for any absolute result,
you must perform some calculations using ->days, which is the only result to be absolute. Even though these calculations
are simple, rewriting them on each of your project is boring and can lead to inattention mistakes.

That's why this library has been made for : to abstract the calculations and to get a more object-oriented coding.

### Get delays using DateTimeInterval
```php
$firstDate = new DateTimeImmutable('2022-07-15');
$secondDate = new DateTimeImmutable('2022-08-15');
$dateTimeInterval = new DateTimeInterval($firstDate, $secondDate);

// Absolute count of days
$dateTimeInterval->getDays();

// Relative count of days
$dateTimeInterval->getDays(false);

// Absolute count of months
$dateTimeInterval->getMonths();

// Relative count of months
$dateTimeInterval->getMonths(false);

// And so on...

// Nostalgic about DateInterval ?
$dateInterval = $dateTimeInterval->getDateInterval();
```

## Copyright and License

The GlucNAc/DateTimeInterval library is copyright © [GlucNAc](https://gitlab.com/GlucNAc) and licensed for use under the
MIT License (MIT). Please see [LICENSE](https://gitlab.com/GlucNAc/DateTimeInterval/-/blob/master/LICENSE) for more information.


[composer]: http://getcomposer.org/

[badge-source]: http://img.shields.io/badge/source-GlucNAc/DateTimeInterval-blue.svg?style=flat-square
[badge-release]: https://img.shields.io/packagist/v/GlucNAc/DateTimeInterval.svg?style=flat-square&label=release
[badge-license]: https://img.shields.io/packagist/l/GlucNAc/DateTimeInterval.svg?style=flat-square
[badge-php]: https://img.shields.io/packagist/php-v/GlucNAc/DateTimeInterval.svg?style=flat-square
[badge-pipeline]: https://gitlab.com/GlucNAc/DateTimeInterval/badges/master/pipeline.svg
[badge-coverage]: https://gitlab.com/GlucNAc/DateTimeInterval/badges/master/coverage.svg
[badge-downloads]: https://img.shields.io/packagist/dt/GlucNAc/DateTimeInterval.svg?style=flat-square&colorB=mediumvioletred

[php]: https://php.net
[source]: https://gitlab.com/GlucNAc/DateTimeInterval
[packagist]: https://packagist.org/packages/GlucNAc/DateTimeInterval
[pipeline]: https://gitlab.com/GlucNAc/DateTimeInterval/-/commits/master
[coverage]: https://gitlab.com/GlucNAc/DateTimeInterval/-/commits/master
[downloads]: https://packagist.org/packages/GlucNAc/DateTimeInterval


CHANGELOG.md
# GlucNAc/DateTimeInterval Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.2] - 2022-07-16

### Changed

* DateTimeInterval (phpDoc)
* README.md

## [2.0.1] - 2022-07-16

### Changed

* Upgrade to php 8.1
* DateTimeInterval (remove phpstan exception)
* README.md

## [2.0.0] - 2022-07-16

### Changed

* DateTimeInterval (upgrade to php 7.4)
* README.md

## [1.0.1] - 2022-07-14

### Added

* Changelog
* .gitlab-ci.yml

### Changed

* Code cleaning to satisfy code sniffers
* README.md

### Removed

* Dockerfile (use [Dockker GlucNAc repo](https://hub.docker.com/repository/docker/glucnac/php) instead)
* travis.yml

## [1.0.0] - 2022-07-10

### Added

* Library creation

[1.0.1]: https://gitlab.com/GlucNAc/DateTimeInterval/-/tags/v1.0.1
[1.0.0]: https://gitlab.com/GlucNAc/DateTimeInterval/-/tags/v1.0.0


tests/DateTimeIntervalTest.php
<?php

namespace GlucNAc\DateTimeInterval\Test;

use DateTime;
use GlucNAc\DateTimeInterval\DateTimeInterval;
use Mockery\Adapter\Phpunit\MockeryTestCase;

class DateTimeIntervalTest extends MockeryTestCase
{
    /*
     * IS NEGATIVE
     */

    /**
     * @covers \GlucNAc\DateTimeInterval\DateTimeInterval::isNegative()
     */
    public function testIsNegativeOk(): void
    {
        $supDate = new DateTime('2022-08-29');
        $infDate = new DateTime('2021-08-28');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate, false);

        $this->assertEquals(true, $dateTimeInterval->isNegative());
    }

    public function testIsNegativeKo(): void
    {
        $supDate = new DateTime('2022-08-29');
        $infDate = new DateTime('2021-08-28');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate, false);

        $this->assertEquals(false, $dateTimeInterval->isNegative());
    }

    public function testIsNegativeForceKo(): void
    {
        $supDate = new DateTime('2022-08-29');
        $infDate = new DateTime('2021-08-28');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate, true);

        $this->assertEquals(false, $dateTimeInterval->isNegative());
    }

    /*
     * YEARS
     */

    public function testGetYearsPositive(): void
    {
        $supDate = new DateTime('2022-08-29');
        $infDate = new DateTime('2021-08-28');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(1, $dateTimeInterval->getYears());
    }

    public function testGetYearsNegative(): void
    {
        $supDate = new DateTime('2022-08-29');
        $infDate = new DateTime('2021-08-28');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-1, $dateTimeInterval->getYears());
    }

    /*
     * MONTHS
     */

    public function testGetMonthsAbsolutePositive(): void
    {
        $supDate = new DateTime('2022-09-28');
        $infDate = new DateTime('2021-06-28');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(15, $dateTimeInterval->getMonths(true));
    }

    public function testGetMonthsRelativePositive(): void
    {
        $supDate = new DateTime('2022-09-28');
        $infDate = new DateTime('2021-06-28');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(3, $dateTimeInterval->getMonths(false));
    }

    public function testGetMonthsAbsoluteNegative(): void
    {
        $supDate = new DateTime('2022-09-28');
        $infDate = new DateTime('2021-06-28');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-15, $dateTimeInterval->getMonths(true));
    }

    public function testGetMonthsRelativeNegative(): void
    {
        $supDate = new DateTime('2022-09-28');
        $infDate = new DateTime('2021-06-28');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-3, $dateTimeInterval->getMonths(false));
    }

    /*
     * DAYS
     */

    public function testGetDaysAbsolutePositive(): void
    {
        $supDate = new DateTime('2021-08-28');
        $infDate = new DateTime('2021-06-29');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(60, $dateTimeInterval->getDays(true));
    }

    public function testGetDaysRelativePositive(): void
    {
        $supDate = new DateTime('2021-08-28');
        $infDate = new DateTime('2021-06-29');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(30, $dateTimeInterval->getDays(false));
    }

    public function testGetDaysAbsoluteNegative(): void
    {
        $supDate = new DateTime('2021-08-28');
        $infDate = new DateTime('2021-06-27');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-62, $dateTimeInterval->getDays(true));
    }

    public function testGetDaysRelativeNegative(): void
    {
        $supDate = new DateTime('2021-08-28');
        $infDate = new DateTime('2021-06-27');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-1, $dateTimeInterval->getDays(false));
    }

    /*
     * HOURS
     */

    public function testGetHoursAbsolutePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:00:00');
        $infDate = new DateTime('2021-08-27 00:00:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(24, $dateTimeInterval->getHours(true));
    }

    public function testGetHoursRelativePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:00:00');
        $infDate = new DateTime('2021-08-27 00:00:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(0, $dateTimeInterval->getHours(false));
    }

    public function testGetHoursAbsoluteNegative(): void
    {
        $infDate = new DateTime('2021-08-27 00:00:00');
        $supDate = new DateTime('2021-08-28 00:00:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-24, $dateTimeInterval->getHours(true));
    }

    public function testGetHoursRelativeNegative(): void
    {
        $infDate = new DateTime('2021-08-27 00:00:00');
        $supDate = new DateTime('2021-08-28 00:00:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(0, $dateTimeInterval->getHours(false));
    }

    /*
     * MINUTES
     */

    public function testGetMinutesAbsolutePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:01:00');
        $infDate = new DateTime('2021-08-27 00:00:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(1441, $dateTimeInterval->getMinutes(true));
    }

    public function testGetMinutesRelativePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:01:00');
        $infDate = new DateTime('2021-08-27 00:00:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(1, $dateTimeInterval->getMinutes(false));
    }

    public function testGetMinutesAbsoluteNegative(): void
    {
        $infDate = new DateTime('2021-08-27 00:00:00');
        $supDate = new DateTime('2021-08-28 00:01:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-1441, $dateTimeInterval->getMinutes(true));
    }

    public function testGetMinutesRelativeNegative(): void
    {
        $infDate = new DateTime('2021-08-27 00:00:00');
        $supDate = new DateTime('2021-08-28 00:01:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-1, $dateTimeInterval->getMinutes(false));
    }

    /*
     * SECONDS
     */

    public function testGetSecondsAbsolutePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:01:00');
        $infDate = new DateTime('2021-08-27 23:59:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(120, $dateTimeInterval->getSeconds(true));
    }

    public function testGetSecondsRelativePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:01:00');
        $infDate = new DateTime('2021-08-27 00:00:00');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(0, $dateTimeInterval->getSeconds(false));
    }

    public function testGetSecondsAbsoluteNegative(): void
    {
        $infDate = new DateTime('2021-08-27 23:59:00');
        $supDate = new DateTime('2021-08-28 00:01:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-120, $dateTimeInterval->getSeconds(true));
    }

    public function testGetSecondsRelativeNegative(): void
    {
        $infDate = new DateTime('2021-08-27 23:59:00');
        $supDate = new DateTime('2021-08-28 00:01:00');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(0, $dateTimeInterval->getSeconds(false));
    }

    /*
     * MICROSECONDS
     */

    public function testGetMicroSecondsAbsolutePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:00:01.100');
        $infDate = new DateTime('2021-08-27 23:59:59.100');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(2000.0, $dateTimeInterval->getMicroSeconds(true));
    }

    public function testGetMicroSecondsRelativePositive(): void
    {
        $supDate = new DateTime('2021-08-28 00:00:01.100');
        $infDate = new DateTime('2021-08-27 23:59:59.100');

        $dateTimeInterval = new DateTimeInterval($supDate, $infDate);

        $this->assertEquals(0.0, $dateTimeInterval->getMicroSeconds(false));
    }

    public function testGetMicroSecondsAbsoluteNegative(): void
    {
        $infDate = new DateTime('2021-08-27 23:59:59.100');
        $supDate = new DateTime('2021-08-28 00:00:01.100');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-2000.0, $dateTimeInterval->getMicroSeconds(true));
    }

    public function testGetMicroSecondsRelativeNegative(): void
    {
        $infDate = new DateTime('2021-08-27 23:59:59.100');
        $supDate = new DateTime('2021-08-28 00:00:01.100');

        $dateTimeInterval = new DateTimeInterval($infDate, $supDate);

        $this->assertEquals(-0.0, $dateTimeInterval->getMicroSeconds(false));
    }
}


.php_cs.php
<?php

$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__ . '/src')
    ->in(__DIR__ . '/tests')
;

$config = new PhpCsFixer\Config();

return $config
    ->setRiskyAllowed(true)
    ->setRules([
        '@Symfony' => true,
        'binary_operator_spaces' => array(
            'align_equals' => false,
            'align_double_arrow' => true,
        ),
        '@Symfony:risky' => true,
        'array_syntax' => ['syntax' => 'short'],
        'linebreak_after_opening_tag' => true,
        'mb_str_functions' => true,
        'no_php4_constructor' => true,
        'no_unreachable_default_argument_value' => true,
        'no_useless_else' => true,
        'no_useless_return' => true,
        'ordered_imports' => true,
        'php_unit_strict' => true,
        'phpdoc_order' => true,
        'semicolon_after_instruction' => true,
        'strict_comparison' => true,
        'strict_param' => true,
        'concat_space' => ['spacing' => 'one'],
        'trailing_comma_in_multiline_array' => true
    ])
    ->setFinder($finder)
    ->setCacheFile(__DIR__.'/.php_cs.cache');

src/DateTimeInterval.php
<?php

namespace GlucNAc\DateTimeInterval;

use DateInterval;
use RuntimeException;

/**
 * Class DateTimeInterval
 *
 * * <em>$supDate</em> should be greater than <em>$infDate</em>.
 *
 * * Returns negative value if <em>$supDate</em> &lt; <em>$infDate</em>,
 * unless <em>$forceReturningPositiveValue</em> = true.
 */
class DateTimeInterval
{
    private DateInterval $dateInterval;

    public function __construct(
        \DateTimeInterface $supDate,
        \DateTimeInterface $infDate,
        bool $returnAbsoluteValue = false
    ) {
        $this->dateInterval = \date_diff($infDate, $supDate, $returnAbsoluteValue);
    }

    /**
     * Returns the count of COMPLETED years between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2022-08-29 and 2021-08-28: returns 1
     *
     * e.g. For 2022-08-28 and 2021-08-28: returns 1
     *
     * e.g. For 2022-08-27 and 2021-08-28: returns 0
     *
     * e.g. For 2022-08-29 01:00:00 and 2021-08-28 00:59:59: returns 0 (period not completed)
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-08-28 and 2022-08-29: returns -1
     *
     * e.g. For 2021-08-28 and 2022-08-28: returns -1
     *
     * e.g. For 2021-08-28 and 2022-08-27: returns 0
     *
     * @see date_diff()
     */
    public function getYears(): int
    {
        $count = $this->dateInterval->y;

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED months between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2022-09-28 and 2021-06-28:
     * * with <em>$absoluteCount</em> = true, this method will return 15
     * * with <em>$absoluteCount</em> = false, this method will return 3
     *
     * e.g. For 2022-09-28 and 2021-06-29:
     * * with <em>$absoluteCount</em> = true, this method will return 14
     * * with <em>$absoluteCount</em> = false, this method will return 2
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2022-06-28 and 2021-09-28:
     * * with <em>$absoluteCount</em> = true, this method will return -15
     * * with <em>$absoluteCount</em> = false, this method will return -3
     *
     * e.g. For 2021-06-29 and 2022-09-28:
     * * with <em>$absoluteCount</em> = true, this method will return -14
     * * with <em>$absoluteCount</em> = false, this method will return -2
     *
     * @see date_diff()
     */
    public function getMonths(bool $absoluteCount = true): int
    {
        $count = $this->dateInterval->m;

        if ($absoluteCount) {
            $count += 12 * $this->dateInterval->y;
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED days between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2021-08-28 and 2021-06-29:
     * * with <em>$absoluteCount</em> = true, this method will return 60
     * * with <em>$absoluteCount</em> = false, this method will return 30
     *
     * e.g. For 2021-08-28 and 2021-06-27:
     * * with <em>$absoluteCount</em> = true, this method will return 62
     * * with <em>$absoluteCount</em> = false, this method will return 1
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-06-27 and 2021-08-28:
     * * with <em>$absoluteCount</em> = true, this method will return -62
     * * with <em>$absoluteCount</em> = false, this method will return -1
     *
     * e.g. For 2021-06-29 and 2021-08-28:
     * * with <em>$absoluteCount</em> = true, this method will return -60
     * * with <em>$absoluteCount</em> = false, this method will return -29
     *
     * e.g. For 2021-06-29 and 2022-08-28:
     * * with <em>$absoluteCount</em> = true, this method will return -425
     * * with <em>$absoluteCount</em> = false, this method will return -29
     *
     * @see date_diff()
     *
     * @throws RuntimeException
     */
    public function getDays(bool $absoluteCount = true): int
    {
        $count = $absoluteCount ? $this->dateInterval->days : $this->dateInterval->d;

        if ($count === false) {
            throw new \RuntimeException('count is false');
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED hours between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2021-08-28 00:00:00 and 2021-08-27 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return 24
     * * with <em>$absoluteCount</em> = false, this method will return 0
     *
     * e.g. For 2021-08-28 01:00:00 and 2021-08-27 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return 25
     * * with <em>$absoluteCount</em> = false, this method will return 1
     *
     * e.g. For 2021-08-28 01:31:00 and 2021-08-27 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return 25
     * * with <em>$absoluteCount</em> = false, this method will return 1
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-08-27 00:00:00 and 2021-08-28 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return -24
     * * with <em>$absoluteCount</em> = false, this method will return 0
     *
     * e.g. For 2021-08-27 00:00:00 and 2021-08-28 01:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return -25
     * * with <em>$absoluteCount</em> = false, this method will return -1
     *
     * @see date_diff()
     */
    public function getHours(bool $absoluteCount = true): int
    {
        $count = $this->dateInterval->h;

        if ($absoluteCount) {
            $count += 24 * $this->dateInterval->days;
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED minutes between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2021-08-28 00:01:00 and 2021-08-27 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return 1441
     * * with <em>$absoluteCount</em> = false, this method will return 1
     *
     * e.g. For 2021-08-28 00:01:59 and 2021-08-27 00:00:00:
     * * with <em>$absoluteCount</em> = true, this method will return 1441
     * * with <em>$absoluteCount</em> = false, this method will return 1
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-08-27 00:00:00 and 2021-08-28 00:01:00:
     * * with <em>$absoluteCount</em> = true, this method will return -1441
     * * with <em>$absoluteCount</em> = false, this method will return -1
     *
     * e.g. For 2021-08-27 00:00:00 and 2021-08-28 00:01:59:
     * * with <em>$absoluteCount</em> = true, this method will return -1441
     * * with <em>$absoluteCount</em> = false, this method will return -1
     *
     * @see date_diff()
     */
    public function getMinutes(bool $absoluteCount = true): int
    {
        $count = $this->dateInterval->i;

        if ($absoluteCount) {
            $count += 60 * abs($this->getHours());
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED seconds between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2021-08-28 00:01:00 and 2021-08-27 23:59:00:
     * * with <em>$absoluteCount</em> = true, this method will return 120
     * * with <em>$absoluteCount</em> = false, this method will return 0
     *
     * e.g. For 2021-08-28 00:01:59 and 2021-08-27 23:59:05:
     * * with <em>$absoluteCount</em> = true, this method will return 115
     * * with <em>$absoluteCount</em> = false, this method will return 55
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-08-27 23:59:00 and 2021-08-28 00:01:00:
     * * with <em>$absoluteCount</em> = true, this method will return -120
     * * with <em>$absoluteCount</em> = false, this method will return 0
     *
     * e.g. For 2021-08-27 23:59:05 and 2021-08-28 00:01:59:
     * * with <em>$absoluteCount</em> = true, this method will return -115
     * * with <em>$absoluteCount</em> = false, this method will return -55
     *
     * @see date_diff()
     */
    public function getSeconds(bool $absoluteCount = true): int
    {
        $count = $this->dateInterval->s;

        if ($absoluteCount) {
            $count += 60 * abs($this->getMinutes());
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * Returns the count of COMPLETED microseconds between the two DateTime instances of this interval.
     *
     * <u>POSITIVE VALUES</u>
     *
     * e.g. For 2021-08-28 00:00:01.100 and 2021-08-27 23:59:59.100:
     * * with <em>$absoluteCount</em> = true, this method will return 2000.0
     * * with <em>$absoluteCount</em> = false, this method will return 0.0
     *
     * e.g. For 2021-08-28 00:00:01.100 and 2021-08-27 23:59:59.000:
     * * with <em>$absoluteCount</em> = true, this method will return 2000.1
     * * with <em>$absoluteCount</em> = false, this method will return 0.1
     *
     * <u>NEGATIVE VALUES</u>
     *
     * e.g. For 2021-08-27 23:59:59.100 and 2021-08-28 00:00:01.100:
     * * with <em>$absoluteCount</em> = true, this method will return -2000.0
     * * with <em>$absoluteCount</em> = false, this method will return -0.0
     *
     * e.g. For 2021-08-27 23:59:59.000 and 2021-08-28 00:00:01.100:
     * * with <em>$absoluteCount</em> = true, this method will return -1999.9
     * * with <em>$absoluteCount</em> = false, this method will return -0.1
     *
     * @see date_diff()
     */
    public function getMicroSeconds(bool $absoluteCount = true): float
    {
        $count = $this->dateInterval->f;

        if ($absoluteCount) {
            $count += 1000 * abs($this->getSeconds());
        }

        return $this->isNegative() ? - $count : $count;
    }

    /**
     * @return bool
     */
    public function isNegative(): bool
    {
        return (bool)$this->dateInterval->invert;
    }

    /**
     * @see DateTimeInterval::format() for accepted format.
     *
     * @codeCoverageIgnore
     */
    public function format(string $format): string
    {
        return $this->dateInterval->format($format);
    }

    /**
     * @return DateInterval|false
     *
     * @codeCoverageIgnore
     */
    public function getDateInterval()
    {
        return $this->dateInterval;
    }
}
